js/flexbuffers.js

```diff
@@ -0,0 +1,1015 @@
+  VECTOR_FLOAT3: 21,
+  VECTOR_INT4: 22,
+  VECTOR_UINT4: 23,
+  VECTOR_FLOAT4: 24,
+  BLOB: 25,
+  BOOL: 26,
+  VECTOR_BOOL: 36,
+};
+
+flexbuffers.ValueType.isInline = (value) => {
+  return value === flexbuffers.ValueType.BOOL
+    || value <= flexbuffers.ValueType.FLOAT;
+};
+
+flexbuffers.ValueType.isNumber = (value) => {
+  return value >= flexbuffers.ValueType.INT
+    && value <= flexbuffers.ValueType.FLOAT;
+};
+
+flexbuffers.ValueType.isIndirectNumber = (value) => {
+  return value >= flexbuffers.ValueType.INDIRECT_INT
+    && value <= flexbuffers.ValueType.INDIRECT_FLOAT;
+};
+
+flexbuffers.ValueType.isTypedVectorElement = (value) => {
+  return value === flexbuffers.ValueType.BOOL
+    || (value >= flexbuffers.ValueType.INT
+      && value <= flexbuffers.ValueType.STRING);
+};
+
+flexbuffers.ValueType.isTypedVector = (value) => {
+  return value === flexbuffers.ValueType.VECTOR_BOOL
+    || (value >= flexbuffers.ValueType.VECTOR_INT
+      && value <= flexbuffers.ValueType.VECTOR_STRING_DEPRECATED);
+};
+
+flexbuffers.ValueType.isFixedTypedVector = (value) => {
+  return value >= flexbuffers.ValueType.VECTOR_INT2
+    && value <= flexbuffers.ValueType.VECTOR_FLOAT4;
+};
+
+flexbuffers.ValueType.isAVector = (value) => {
+  return flexbuffers.ValueType.isTypedVector(value)
+    || flexbuffers.ValueType.isFixedTypedVector(value)
+    || value === flexbuffers.ValueType.VECTOR;
+};
+
+flexbuffers.ValueType.toTypedVector = (valueType, length) => {
+  if (length === 0) return valueType - flexbuffers.ValueType.INT + flexbuffers.ValueType.VECTOR_INT;
+  if (length === 2) return valueType - flexbuffers.ValueType.INT + flexbuffers.ValueType.VECTOR_INT2;
+  if (length === 3) return valueType - flexbuffers.ValueType.INT + flexbuffers.ValueType.VECTOR_INT3;
+  if (length === 4) return valueType - flexbuffers.ValueType.INT + flexbuffers.ValueType.VECTOR_INT4;
+  throw "Unexpected length " + length;
+};
+
+flexbuffers.ValueType.typedVectorElementType = (valueType) => {
+  return valueType - flexbuffers.ValueType.VECTOR_INT + flexbuffers.ValueType.INT;
+};
+
+flexbuffers.ValueType.fixedTypedVectorElementType = (valueType) => {
+  return ((valueType - flexbuffers.ValueType.VECTOR_INT2) % 3) + flexbuffers.ValueType.INT;
+};
+
+flexbuffers.ValueType.fixedTypedVectorElementSize = (valueType) => {
+  // The x / y >> 0 trick is to have an int division. Suppose to be faster than Math.floor()
+  return (((valueType - flexbuffers.ValueType.VECTOR_INT2) / 3) >> 0)  + 2;
+};
+
+flexbuffers.ValueType.packedType = (valueType, bitWidth) => {
+  return bitWidth | (valueType << 2);
+};
+
+flexbuffers.toReference = (buffer) => {
+
+  function validateOffset(dataView, offset, width) {
+    if (dataView.byteLength <= offset + width || offset & (flexbuffers.BitWidth.toByteWidth(width) - 1) !== 0) {
+      throw "Bad offset: " + offset + ", width: " + width;
+    }
+  }
+
+  function readInt(dataView, offset, width) {
+    validateOffset(dataView, offset, width);
```

Comment:
The other implementations do not have this offset validations, it be much better for speed to simply rely on whatever bounds checking the buffer this sits in does.

Reply:
Removed the call, kept the function in case someone needs to debug. Would love to have real conditional compilation in JS :)